### ISSUE TITLE
Changes pagination some requests returning a tuple from previous request

### ISSUE THREAD

**Issue by collisj1996** [openfga/openfga#1078]
**Summary**

Testing pagination on the changes endpoint I chose a page size of 2, some of the subsequent requests would have duplicate values from the previous request

- Steps to reproduce
  - Using the Python openfga_sdk
  - Using a page size of 2
  - Using 5 tuples (see below)
  - calling the change endpoint 3 times in succession each using the previous requests continuation token
  - The second request has a duplicate tuple
  - This results in 6 tuples in total with one duplicate
  
- What did you expect to happen?

That each request using the previous request continuation token would have newer distinct tuples

- What actually happened?

Some subsequent request using the previous requests continuation token would include the duplicate values

**OpenFGA version or commit**

Latest OpenFga docker image as of 24th Oct

**Store data**

If applicable, provide information about your authorization model, tuples, and the Check or ListObjects calls you're making. For example:

```yaml
model_file: |
model
  schema 1.1
type user
type domain
  relations
    define member: [user] or admin
    define admin: [user]
    define can_create_folder: member or admin
    define can_create_playlist: member or admin
    define can_view_item_type: member or admin
    define can_view_tag: member or admin
    define can_check_action: member or admin
    define can_create_item_type: admin
    define can_delete_item_type: admin
    define can_rename_item_type: admin
    define can_view_role: admin
    define can_edit_role: admin
    define can_create_tag: admin
    define can_update_tag: admin
    define can_delete_tag: admin
type playlist
  relations
    define can_view: base_can_view and member from domain
    define can_update_description: base_can_update_description and member from domain
    define can_delete: base_can_delete and member from domain
    define can_change_item_type: base_can_change_item_type and member from domain
    define can_rename: base_can_rename and member from domain
    define can_update_items: base_can_update_items and member from domain
    define base_can_view: view from policy or owner
    define base_can_update_description: edit from policy or owner
    define base_can_delete: delete from policy or owner
    define base_can_change_item_type: admin from policy or owner
    define base_can_rename: admin from policy or owner
    define base_can_update_items: admin from policy or owner
    define policy: [policy]
    define owner: [user]
    define domain: [domain]
type folder
  relations
    define can_view_in_list: base_can_view_in_list and member from domain
    define can_view: base_can_view and member from domain
    define can_edit_content: base_can_edit_content and member from domain
    define can_delete: base_can_delete and member from domain
    define can_change_item_type: base_can_change_item_type and member from domain
    define can_move: base_can_move and member from domain
    define can_rename: base_can_rename and member from domain
    define base_can_view_in_list: any from policy or owner
    define base_can_view: view from policy or owner
    define base_can_edit_content: edit from policy or owner
    define base_can_delete: delete from policy or owner
    define base_can_change_item_type: admin from policy or owner
    define base_can_move: admin from policy or owner
    define base_can_rename: admin from policy or owner
    define policy: [policy]
    define owner: [user]
    define parent: [folder]
    define domain: [domain]
type file
  relations
    define can_view_in_list: base_can_view_in_list and member from domain
    define can_view: base_can_view and member from domain
    define can_view_notes: base_can_view_notes and member from domain
    define can_add_notes: base_can_add_notes and member from domain
    define can_modify_notes: base_can_modify_notes and member from domain
    define can_delete_notes: base_can_delete_notes and member from domain
    define can_export_notes: base_can_export_notes and member from domain
    define can_update_description: base_can_update_description and member from domain
    define can_create_subclip: base_can_create_subclip and member from domain
    define can_download_clear: base_can_download_clear and member from domain
    define can_download_restricted: base_can_download_restricted and member from domain
    define can_delete: base_can_delete and member from domain
    define can_change_item_type: base_can_change_item_type and member from domain
    define can_rename: base_can_rename and member from domain
    define can_move: base_can_move and member from domain
    define can_add_asset: base_can_add_asset and member from domain
    define can_refresh_upload: base_can_refresh_upload and member from domain
    define can_publish_checksums: base_can_publish_checksums and member from domain
    define base_can_view_in_list: any from policy or any from item_type or owner
    define base_can_view: view from policy or view from item_type or owner
    define base_can_view_notes: view from policy or view from item_type or owner
    define base_can_add_notes: view from policy or view from item_type or owner
    define base_can_modify_notes: view from policy or view from item_type or owner
    define base_can_delete_notes: view from policy or view from item_type or owner
    define base_can_export_notes: view from policy or view from item_type or owner
    define base_can_update_description: edit from policy or edit from item_type or owner
    define base_can_create_subclip: edit from policy or edit from item_type or owner
    define base_can_download_clear: download_clear from policy or download_clear from item_type or owner
    define base_can_download_restricted: download_restricted from policy or download_restricted from item_type or owner
    define base_can_delete: delete from policy or delete from item_type or owner
    define base_can_change_item_type: admin from policy or admin from item_type or owner
    define base_can_rename: admin from policy or admin from item_type or owner
    define base_can_move: admin from policy or admin from item_type or owner
    define base_can_add_asset: admin from policy or admin from item_type or owner
    define base_can_refresh_upload: admin from policy or admin from item_type or owner
    define base_can_publish_checksums: admin from policy or admin from item_type or owner
    define policy: [policy]
    define owner: [user]
    define parent: [folder]
    define item_type: [item_type]
    define domain: [domain]
type role
  relations
    define assignee: [user,domain#member]
type item_type
  relations
    define view: [role#assignee]
    define edit: [role#assignee]
    define download_clear: [role#assignee]
    define download_restricted: [role#assignee]
    define delete: [role#assignee]
    define share: [role#assignee]
    define admin: [role#assignee]
    define any: view or edit or download_clear or download_restricted or delete or share or admin
type policy
  relations
    define view: [user,domain#member]
    define edit: [user,domain#member]
    define download_clear: [user,domain#member]
    define download_restricted: [user,domain#member]
    define delete: [user,domain#member]
    define share: [user,domain#member]
    define admin: [user,domain#member]
    define any: view or edit or download_clear or download_restricted or delete or share or admin
    
tuples:
  - user: user:user_one
    relation: owner
    object: file:file_one
  - user: user:user_one
    relation: owner
    object: file:file_two
  - user: user:user_two
    relation: owner
    object: file:file_three
  - user: user:user_one
    relation: admin
    object: domain:domain_one
  - user: user:user_two
    relation: admin
    object: domain:domain_one
```

**Other data**

- NOTE:

  - The tuples are being added transactionally in one write request, I'm not sure if this is relevant but might cause a bug to the pagination logic when fetching??

- How are you running OpenFGA?

  - Docker (latest image)

- What datastore are you using? (In memory, MySQL, Postgres)

  - Postgres

- Are you running OpenFGA with any configuration overrides or with any of the flags mentioned in `./openfga run --help`?

  - No

- Do you have any logs or traces that could help us debug the problem?

  - Not that I know of

**Comment by miparnisari** [comment]
Hi @collisj1996, thanks for the report. I was able to reproduce locally. What's interesting is that we have tests for a similar scenario [here](https://github.com/openfga/openfga/blob/56ea59e9abc568f2cc5c6a6cfd8451fb87daca9c/pkg/server/test/read_changes.go#L72) but these are passing. We need to dig deeper to find if there are any differences between the test and the data you provided. Maybe the test is incorrect 🤔

**Comment by collisj1996** [comment]
I'm not sure how your internal pagination works, I'm not overly familiar with this codebase, however, I did notice in this example you're only ordering by `inserted_at`, but because they were added transactionally they have the same timestamp;

to have predictable ordering should we not order by timestamp first and then the tuple itself `(file:file_one,owner,user:user_one)` at least in this way the ordering is predictable

**Comment by miparnisari** [comment]
> I did notice in this example you're only ordering by inserted_at, but because they were added transactionally they have the same timestamp;

I forget why we did it like this (or if we even discussed it). Like you said, `inserted_at` is not guaranteed to be unique. This seems to be similar to https://stackoverflow.com/questions/19304579/postgresql-limit-order-returning-inconsistent-results

I think we should we shorting the changes by `ulid`.

>  I also wonder if this could be a Postgres collation issue

I'm not sure, could be.

**Comment by miparnisari** [comment]
Fixed in 1.3.5. Thanks @collisj1996 for the report!
